src/ninja.cc

```diff
@@ -1087,6 +1096,124 @@ int NinjaMain::ToolRestat(const Options* options, int argc, char* argv[]) {
   return EXIT_SUCCESS;
 }
 
+struct Compdb_Targets {
+  enum class Action {
+    display_help_and_exit,
+    emit_commands
+  };
+
+  Action action;
+  EvaluateCommandMode eval_mode = ECM_NORMAL;
+
+  std::vector<StringPiece> targets;
+
+  static Compdb_Targets parse(int argc, char* argv[]) {
+    //
+    // grammar:
+    //     ninja -t compdb-targets [-hx] target [targets]
+    //
+    Compdb_Targets ret;
+
+    // getopt_long() expects argv[0] to contain the name of
+    // the tool, i.e. "compdb".
+    argc++;
+    argv--;
+
+    // Phase 1: parse options:
+    {
+      optind = 1;  // see `man 3 getopt` for documentation on optind
+      int opt;
+      while ((opt = getopt(argc, argv, "hx")) != -1) {
+        switch (opt) {
+        case 'x':
+          ret.eval_mode = ECM_EXPAND_RSPFILE;
+          break;
+        case 'h':
+        default:
+          ret.action = Compdb_Targets::Action::display_help_and_exit;
+          goto fin;
+        }
+      }
+    }
+
+    // Phase 2: parse operands:
+    {
+      int targets_begin = optind;
+      int const targets_end = argc;
+
+      if (targets_begin == targets_end) {
+        Fatal("compdb-targets expects the name of at least one target");
+      } else {  // parse rules & targets (if any)
+        ret.action = Compdb_Targets::Action::emit_commands;
+        for (int i = targets_begin; i < targets_end; ++i)
+          ret.targets.push_back(argv[i]);
+      }
+    }
+
+  fin:
+    return ret;
+  }
+};
+
+
+void printCompdb(std::string const &directory, std::vector<Edge *> const &edges,
+                 const EvaluateCommandMode eval_mode) {
+  putchar('[');
+
+  bool first = true;
+  for (Edge* edge : edges) {
+    if (edge->is_phony() || edge->inputs_.empty())
+      continue;
+    if (!first)
+      putchar(',');
+    printOneCompdbObject(directory, edge, eval_mode);
+    first = false;
+  }
+
+  puts("\n]");
+}
+
+int NinjaMain::ToolCompilationDatabaseForTargets(const Options* options,
+                                                 int argc, char* argv[]) {
+  auto compdb = Compdb_Targets::parse(argc, argv);
+
+  switch (compdb.action) {
+  case Compdb_Targets::Action::display_help_and_exit: {
+    printf(
+        "usage: ninja -t compdb [-hx] target [targets]\n"
+        "\n"
+        "options:\n"
+        "  -h     display this help messsage\n"
+        "  -x     expand @rspfile style response file invocations\n");
+    return 1;
+  }
+
+  case Compdb_Targets::Action::emit_commands: {
+    InEdge_Collector inEdge_collector;
```

Comment:
Avoid mixing CamelCase and snake_case whenever possible. Variable names use snake_case. So `in_edge_collector` or even just `edge_collector` or `collector` would be fine here.

Reply:
fwiw, i deliberately mix camel case and snake case whenever the equivalent english phrase would be partially hyphenated; e.g. an "in-edge collector" is a collector that collects the (direct and indirect) in-edges of the given node (as opposed to out-edges of said node), so called because it uses `node->in_edge()` (as opposed to `node->out_edges()`)

since hyphens are not allowed in C++ identifiers, i squish the hyphenated part into camel case, and replace the space with an underscore; thus "in-edge collector" is transformed into "InEdge_Collector".

in the case of the variable, i'm fine with renaming it `collector`, but i guess the class would have to be renamed to `In_Edge_Collector`, which... feels unfortunate, because it loses the grouping, and on first read, you could mistakenly read it as "in edge-collector" (i.e. "to be inside an edge collector"...?)

But i'm not going to die on this particular hill. (:

Reply:
agreed!
[happily retreats from hill]